{{cookiecutter.project_slug}}/.gitlab-ci.yml

```diff
@@ -0,0 +1,130 @@
+# templates
+# ------------
+
+.deploy-template:
+  image: alpine:latest
+  script:
+    - echo "Deploying to ${CI_ENVIRONMENT_NAME}"
+  variables:
+    GIT_STRATEGY: none  # prevents fetching repo (we only need the artifact here)
+
+# -------------
+# Pipeline Definition
+
+variables:
+  PIP_CACHE_DIR: $CI_PROJECT_DIR/.cache/pip
+  {%- if cookiecutter.package_manager == 'conda' %}
+  CONDA_PKGS_DIRS: $CI_PROJECT_DIR/.cache/conda
+  {%- elif cookiecutter.package_manager == 'poetry' %}
+  POETRY_VERSION: 1.0.5
+  POETRY_CACHE_DIR: $CI_PROJECT_DIR/.cache/poetry
+  {%- endif %}
+
+stages:
+  - build
+  - test
+  - deploy-nonprod
+  - deploy-prod
+
+# Build and Test
+{% if cookiecutter.package_manager == 'poetry' -%}
+build-wheel:
+  image: python:3-slim
+  stage: build
+  artifacts:
+    name: 'app-wheel'
+    paths:
+      - dist/{{ cookiecutter.module_name }}-*.whl
+    expire_in: 6 mos
+  script:
+    - pip install poetry==$POETRY_VERSION
+    - poetry build -f wheel
+{% else -%}
+build-wheel:
+  image: python:3-slim
+  stage: build
+  artifacts:
+    name: 'app-wheel'
+    paths:
+      - dist/{{ cookiecutter.module_name }}-*.whl
+    expire_in: 6 mos
+  script:
+    - python setup.py dist
+{%- endif %}
+
+{% if cookiecutter.package_manager == 'poetry' -%}
+test-unit:
+  stage: test
+  image: python:3-slim
+  cache:
+    key:
+      files:
+        - pyproject.toml
+    paths:
+      - $PIP_CACHE_DIR
+      - $POETRY_CACHE_DIR
+  before_script:
+    - pip install poetry==$POETRY_VERSION
+    - poetry install --no-root
+    - source `poetry env info --path`/bin/activate
+  script:
+    - pytest tests
+{% elif cookiecutter.package_manager == 'conda' -%}
+test-unit:
+  stage: test
+  image: continuumio/miniconda3
+  cache:
+    key:
+      files:
+        - environment.yml
+        - environment-dev.yml
+    paths:
+      - $PIP_CACHE_DIR
+      - $CONDA_PKGS_DIRS/*.tar.bz2
+      - $CONDA_PKGS_DIRS/urls.txt
+  before_script:
+    - export PATH="/opt/conda/bin:$PATH"
+    - conda env create -n .venv -f environment-dev.yml environment.yml
+    - source activate .venv
+    - pip install dist/{{ cookiecutter.module_name }}-*.whl
+  script:
+    - pytest tests
+{% elif cookiecutter.package_manager == 'pip' -%}
+test-unit:
+  stage: test
+  image: python:3-slim
+  cache:
+    key:
+      files:
+        - requirements.txt
+        - requirements-dev.txt
+    paths:
+      - $PIP_CACHE_DIR
```

Comment:
I managed to get the cache working with the virtualenv-folder, so you don't have to install the packages anymore (which can take some time, especially when there are no wheel packages available), but it might still have issues in some edge cases. So this is probably the safer solution for now

Reply:
@klamann Yes, I also tested conda and poetry.